README.md
<p align="center">
  <a href="https://www.algolia.com/doc/guides/building-search-ui/what-is-instantsearch/js/">
    <img alt="InstantSearch.js" src=".github/banner.png">
  </a>

  <p align="center">
    InstantSearch.js is a JavaScript library for building performant and instant search experiences with <a href="https://www.algolia.com/?utm_source=instantsearch.js&utm_campaign=repository">Algolia</a>.
  </p>
</p>

---

[![Version][version-svg]][package-url]
[![License][license-image]][license-url]
[![Build Status][ci-svg]][ci-url]
[![Pull reminders][pull-reminders-svg]][pull-reminders-url]

InstantSearch.js is a vanilla JavaScript library that lets you create an instant-search result experience using [Algolia][algolia-website]’s search API. It is part of the InstantSearch family:

**InstantSearch.js** | [React InstantSearch][react-instantsearch-github] | [Vue InstantSearch][vue-instantsearch-github] | [Angular InstantSearch][instantsearch-angular-github] | [React InstantSearch Native][react-instantsearch-github] | [InstantSearch Android][instantsearch-android-github] | [InstantSearch iOS][instantsearch-ios-github]

<details>
  <summary><strong>Table of contents</strong></summary>

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Why](#why)
- [Getting started](#getting-started)
- [Installation](#installation)
  - [TypeScript users](#typescript-users)
- [Documentation](#documentation)
- [Demos](#demos)
- [Playground](#playground)
- [Browser support](#browser-support)
- [Troubleshooting](#troubleshooting)
- [Contributing](#contributing)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

</details>

## Why

You should be using InstantSearch if you want to:

- Design search experiences with best practices
- Customize your components at will
- Remain independent from external frameworks

## Getting started

Using InstantSearch.js is as simple as adding this JavaScript code to your page:

```javascript
// 1. Instantiate the search
const search = instantsearch({
  indexName: 'instant_search',
  searchClient: algoliasearch('latency', '6be0576ff61c053d5f9a3225e2a90f76'),
});

search.addWidgets([
  // 2. Create an interactive search box
  instantsearch.widgets.searchBox({
    container: '#searchbox',
    placeholder: 'Search for products',
  }),

  // 3. Plug the search results into the product container
  instantsearch.widgets.hits({
    container: '#products',
    templates: {
      item: '{{#helpers.highlight}}{ "attribute": "name" }{{/helpers.highlight}}',
    },
  }),

  // 4. Make the brands refinable
  instantsearch.widgets.refinementList({
    container: '#brand',
    attribute: 'brand',
  }),
]);

// 5. Start the search!
search.start();
```

<p align="center">
  <a href="https://codesandbox.io/s/github/algolia/doc-code-samples/tree/master/InstantSearch.js/getting-started" title="Edit on CodeSandbox">
    <img alt="Edit on CodeSandbox" src="https://codesandbox.io/static/img/play-codesandbox.svg">
  </a>
</p>

To learn more about the library, follow the [getting started](https://www.algolia.com/doc/guides/building-search-ui/getting-started/js/) guide or check how to [add it to your own project](https://www.algolia.com/doc/guides/building-search-ui/installation/js/).

## Installation

```sh
npm install instantsearch.js algoliasearch
# or
yarn add instantsearch.js algoliasearch
```

### TypeScript users

To use InstantSearch.js in a TypeScript environment, depending on your [`algoliasearch`](https://github.com/algolia/algoliasearch-client-javascript) version, you need to import different types.

>You still need to import these types even if you don't use InstantSearch.js with [`algoliasearch`](https://github.com/algolia/algoliasearch-client-javascript).

#### `algoliasearch` v4.x

This version uses types provided by both `algoliasearch` and `@algolia/client-search`.

```bash
yarn add algoliasearch@4 @algolia/client-search
```

#### `algoliasearch` v3.x

```bash
yarn add @types/algoliasearch@3
```

>v3.x is deprecated and will soon no longer be supported.

## Documentation

The documentation is available on the [Algolia website](https://www.algolia.com/doc/guides/building-search-ui/what-is-instantsearch/js/).

## Demos

| E-commerce                                                                                                                                                                                                                   | Media                                                                                                                                                                                                         | Travel                                                                                                                                                                                                              |
| ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| <a href="https://instantsearchjs.netlify.com/examples/e-commerce/"><img src="https://www.algolia.com/doc/assets/images/build-search-ui/demos/e-commerce-2c7ed6b6.png" width="250" alt="E-commerce demo preview"></a> | <a href="https://instantsearchjs.netlify.com/examples/media/"><img src="https://www.algolia.com/doc/assets/images/build-search-ui/demos/media-articles-abc153ab.png" width="250" alt="Media demo preview"></a> | <a href="https://instantsearchjs.netlify.com/examples/tourism/"><img src="https://instantsearchjs.netlify.com/examples/tourism/capture.png" width="250" alt="Tourism demo preview"></a> |

See more demos [on the website](https://www.algolia.com/doc/guides/building-search-ui/resources/demos/js/).

## Playground

You can get to know InstantSearch.js on [this playground](https://codesandbox.io/s/github/algolia/create-instantsearch-app/tree/templates/instantsearch.js).

Start by [adding widgets](https://www.algolia.com/doc/guides/building-search-ui/widgets/showcase/js/) and tweaking the display. Once you feel familiar with the library, we recommend following the [getting started guide](https://www.algolia.com/doc/guides/building-search-ui/getting-started/js/).

## Browser support

We support the **last two versions of major browsers** (Chrome, Edge, Firefox, Safari).

Please refer to the [browser support](https://www.algolia.com/doc/guides/building-search-ui/installation/js/#browser-support) section in the documentation to use InstantSearch.js on other browsers.

## Troubleshooting

Encountering an issue? Before reaching out to support, we recommend heading to our [FAQ](https://www.algolia.com/doc/guides/building-search-ui/troubleshooting/faq/js/) where you will find answers for the most common issues and gotchas with the library.

## Contributing

We welcome all contributors, from casual to regular 💙

- **Bug report**. Is something not working as expected? [Send a bug report](https://github.com/algolia/instantsearch.js/issues/new?template=Bug_report.md).
- **Feature request**. Would you like to add something to the library? [Send a feature request](https://github.com/algolia/instantsearch.js/issues/new?template=Feature_request.md).
- **Documentation**. Did you find a typo in the doc? [Open an issue](https://github.com/algolia/instantsearch.js/issues/new) and we'll take care of it.
- **Development**. If you don't know where to start, you can check the open issues that are [tagged easy](https://github.com/algolia/instantsearch.js/issues?q=is%3Aopen+is%3Aissue+label%3A%22Difficulty%3A++++++%E2%9D%84%EF%B8%8F+easy%22), the [bugs](https://github.com/algolia/instantsearch.js/issues?q=is%3Aissue+is%3Aopen+label%3A%22%E2%9D%A4+Bug%22) or [chores](https://github.com/algolia/instantsearch.js/issues?q=is%3Aissue+is%3Aopen+label%3A%22%E2%9C%A8+Chore%22).

To start contributing to code, you need to:

1.  [Fork the project](https://help.github.com/articles/fork-a-repo/)
1.  [Clone the repository](https://help.github.com/articles/cloning-a-repository/)
1.  Install the dependencies: `yarn`
1.  Run the development mode: `cd packages/instantsearch.js && yarn storybook`
1.  [Open the stories](http://localhost:6006)

Please read [our contribution process](CONTRIBUTING.md) to learn more.

## License

InstantSearch.js is [MIT licensed][license-url].

<!-- Badges -->

[version-svg]: https://img.shields.io/npm/v/instantsearch.js.svg?style=flat-square
[package-url]: https://npmjs.org/package/instantsearch.js
[ci-svg]: https://img.shields.io/circleci/project/github/algolia/instantsearch.js.svg?style=flat-square
[ci-url]: https://circleci.com/gh/algolia/instantsearch.js
[pull-reminders-svg]: https://img.shields.io/badge/pull%20reminders-✓-success.svg?style=flat-square
[pull-reminders-url]: https://pullreminders.com?ref=badge
[license-image]: http://img.shields.io/badge/license-MIT-green.svg?style=flat-square
[license-url]: LICENSE

<!-- Links -->

[algolia-website]: https://www.algolia.com/?utm_source=instantsearch.js&utm_campaign=repository
[react-instantsearch-github]: https://github.com/algolia/react-instantsearch/
[vue-instantsearch-github]: https://github.com/algolia/vue-instantsearch
[instantsearch-android-github]: https://github.com/algolia/instantsearch-android
[instantsearch-ios-github]: https://github.com/algolia/instantsearch-ios
[instantsearch-angular-github]: https://github.com/algolia/angular-instantsearch


CHANGELOG.md
## [4.49.2](https://github.com/algolia/instantsearch.js/compare/v4.49.1...v4.49.2) (2022-11-22)


### Bug Fixes

* **answers:** mark as deprecated ([#5157](https://github.com/algolia/instantsearch.js/issues/5157)) ([d7b62ed](https://github.com/algolia/instantsearch.js/commit/d7b62eda77d894a3677a0967a05a2a357e7e5a94))



## [4.49.1](https://github.com/algolia/instantsearch.js/compare/v4.49.0...v4.49.1) (2022-11-01)


### Bug Fixes

* **insights:** check before usage of `document` ([#5149](https://github.com/algolia/instantsearch.js/issues/5149)) ([6733dea](https://github.com/algolia/instantsearch.js/commit/6733dea1091a3a6c8ec9049eba652a7f06e9c501))



# [4.49.0](https://github.com/algolia/instantsearch.js/compare/v4.48.1...v4.49.0) (2022-10-25)


### Features

* **poweredBy:** update component logo ([#5145](https://github.com/algolia/instantsearch.js/issues/5145)) ([7df7816](https://github.com/algolia/instantsearch.js/commit/7df7816eac1bb3d2eafee5da7b6f4f59611468b2))



## [4.48.1](https://github.com/algolia/instantsearch.js/compare/v4.48.0...v4.48.1) (2022-10-18)


### Bug Fixes

* **bundlesize:** consolidate usage of "classnames" helper ([#5138](https://github.com/algolia/instantsearch.js/issues/5138)) ([f1ec288](https://github.com/algolia/instantsearch.js/commit/f1ec28889be5c2f906dd398f37d072587e29cf3a))
* **currentRefinements:** reset page number on refine ([#5136](https://github.com/algolia/instantsearch.js/issues/5136)) ([407b576](https://github.com/algolia/instantsearch.js/commit/407b5767b51c26d5f471071a92f2e32762898f24))
* **events:** prevent warning on low number of listeners ([#5143](https://github.com/algolia/instantsearch.js/issues/5143)) ([432aa70](https://github.com/algolia/instantsearch.js/commit/432aa7006e7d8eefd1c8c382f59ea2d2974a19da))



# [4.48.0](https://github.com/algolia/instantsearch.js/compare/v4.47.0...v4.48.0) (2022-10-10)


### Bug Fixes

* **insightsMiddleware:** infer type of insightsClient for onEvent ([#5130](https://github.com/algolia/instantsearch.js/issues/5130)) ([dd5fca4](https://github.com/algolia/instantsearch.js/commit/dd5fca4c185c66f1e31ebe9c0568bcad48e062f3)), closes [#5129](https://github.com/algolia/instantsearch.js/issues/5129)


### Features

* **routing:** include repeated indexId in URL correctly ([#5134](https://github.com/algolia/instantsearch.js/issues/5134)) ([679f5da](https://github.com/algolia/instantsearch.js/commit/679f5dad839536def6ae9c3a18416296d40ed49a))



# [4.47.0](https://github.com/algolia/instantsearch.js/compare/v4.46.3...v4.47.0) (2022-10-03)


### Bug Fixes

* **hierarchicalMenu:** pass correct attribute name to Insights ([#5124](https://github.com/algolia/instantsearch.js/issues/5124)) ([fe18a16](https://github.com/algolia/instantsearch.js/commit/fe18a168b1b195d067298770b55fd29a7fdb6edb))


### Features

* **status:** introduce status in InstantSearch class ([#5115](https://github.com/algolia/instantsearch.js/issues/5115)) ([21f3147](https://github.com/algolia/instantsearch.js/commit/21f31476e75e162b38b002d5439f231f3990e785))
* **hierarchicalMenu**: introduce `ais-HierarchicalMenu-item--selected` class ([#5125](https://github.com/algolia/instantsearch.js/issues/5125)) ([4ebb828](https://github.com/algolia/instantsearch.js/commit/4ebb828c93afabfd8083246dfe7edfd33932d5fd))


## [4.46.3](https://github.com/algolia/instantsearch.js/compare/v4.46.2...v4.46.3) (2022-09-27)


### Bug Fixes

* **currentRefinements:** implement noRefinementRoot modifier class ([#5114](https://github.com/algolia/instantsearch.js/issues/5114)) ([cb66830](https://github.com/algolia/instantsearch.js/commit/cb668305af26bf919841c25bd4cc8493fcdf8cf9))



## [4.46.2](https://github.com/algolia/instantsearch.js/compare/v4.46.1...v4.46.2) (2022-09-22)


### Bug Fixes

* **build:** remove jsx pragma comments from build output ([#5112](https://github.com/algolia/instantsearch.js/issues/5112)) ([6582083](https://github.com/algolia/instantsearch.js/commit/65820831b7d7e14867f13a2947795491730b8442))
* **imports:** split out templating from ./utils ([#5111](https://github.com/algolia/instantsearch.js/issues/5111)) ([fc765f3](https://github.com/algolia/instantsearch.js/commit/fc765f35ddd85068237edc81c66932b098e3b55a)), closes [#5109](https://github.com/algolia/instantsearch.js/issues/5109)



## [4.46.1](https://github.com/algolia/instantsearch.js/compare/v4.46.0...v4.46.1) (2022-09-15)


### Bug Fixes

* **hierarchicalMenu:** use existing facet filters in multi queries for parent facet values ([#5105](https://github.com/algolia/instantsearch.js/issues/5105)) ([10a83f1](https://github.com/algolia/instantsearch.js/commit/10a83f146f714d9f97bb8edca2499f16df4ca22d))
* **insights:** make sure change in userToken can't reset the search parameters ([#5101](https://github.com/algolia/instantsearch.js/issues/5101)) ([b20c8dc](https://github.com/algolia/instantsearch.js/commit/b20c8dc70e34c1f234dc10eb7fc69296f30986a4))
* **setUiState**: call onStateChange handler ([#5104](https://github.com/algolia/instantsearch.js/issues/5104)) ([231853d](https://github.com/algolia/instantsearch.js/commit/231853dab731189a33ee480cdb196789c7336fda)))



## [4.46.0](https://github.com/algolia/instantsearch.js/compare/v4.45.1...v4.46.0) (2022-09-12)


### Features

* **html:** deprecate Hogan.js and string-based templates ([#5095](https://github.com/algolia/instantsearch.js/issues/5095)) ([a06ddf5](https://github.com/algolia/instantsearch.js/commit/a06ddf55f1ffd1a93cddab2fcf95d2be3220a423))
* **html:** introduce `html` templating ([#5081](https://github.com/algolia/instantsearch.js/issues/5081)) ([e55e224](https://github.com/algolia/instantsearch.js/commit/e55e2245256193d27f2c85f24b8aab7c9048c554))



## [4.45.1](https://github.com/algolia/instantsearch.js/compare/v4.45.0...v4.45.1) (2022-09-06)


### Bug Fixes


schema.js
'use strict'; // code generated by pbf v3.2.1

var WidgetType = exports.WidgetType = {
    "ais.custom": {
        "value": 0,
        "options": {}
    },
    "ais.instantSearch": {
        "value": 1,
        "options": {}
    },
    "ais.autocomplete": {
        "value": 2,
        "options": {}
    },
    "ais.breadcrumb": {
        "value": 3,
        "options": {}
    },
    "ais.clearRefinements": {
        "value": 4,
        "options": {}
    },
    "ais.configure": {
        "value": 5,
        "options": {}
    },
    "ais.configureRelatedItems": {
        "value": 6,
        "options": {}
    },
    "ais.currentRefinements": {
        "value": 7,
        "options": {}
    },
    "ais.geoSearch": {
        "value": 8,
        "options": {}
    },
    "ais.hierarchicalMenu": {
        "value": 9,
        "options": {}
    },
    "ais.hits": {
        "value": 10,
        "options": {}
    },
    "ais.hitsPerPage": {
        "value": 11,
        "options": {}
    },
    "ais.index": {
        "value": 12,
        "options": {}
    },
    "ais.infiniteHits": {
        "value": 13,
        "options": {}
    },
    "ais.menu": {
        "value": 14,
        "options": {}
    },
    "ais.numericMenu": {
        "value": 15,
        "options": {}
    },
    "ais.pagination": {
        "value": 16,
        "options": {}
    },
    "ais.places": {
        "value": 17,
        "options": {}
    },
    "ais.poweredBy": {
        "value": 18,
        "options": {}
    },
    "ais.queryRules": {
        "value": 19,
        "options": {}
    },
    "ais.queryRuleCustomData": {
        "value": 20,
        "options": {}
    },
    "ais.queryRuleContext": {
        "value": 21,
        "options": {}
    },
    "ais.range": {
        "value": 22,
        "options": {}
    },
    "ais.rangeInput": {
        "value": 23,
        "options": {}
    },
    "ais.rangeSlider": {
        "value": 24,
        "options": {}
    },
    "ais.ratingMenu": {
        "value": 25,
        "options": {}
    },
    "ais.refinementList": {
        "value": 26,
        "options": {}
    },
    "ais.searchBox": {
        "value": 27,
        "options": {}
    },
    "ais.sortBy": {
        "value": 28,
        "options": {}
    },
    "ais.stats": {
        "value": 29,
        "options": {}
    },
    "ais.toggleRefinement": {
        "value": 30,
        "options": {}
    },
    "ais.voiceSearch": {
        "value": 31,
        "options": {}
    }
};

var WidgetParams = exports.WidgetParams = {
    "indexName": {
        "value": 1000,
        "options": {}
    },
    "searchClient": {
        "value": 1001,
        "options": {}
    },
    "numberLocale": {
        "value": 1002,
        "options": {}
    },
    "searchFunction": {
        "value": 1003,
        "options": {}
    },
    "initialUiState": {
        "value": 1004,
        "options": {}
    },
    "onStateChange": {
        "value": 1005,
        "options": {}
    },
    "stalledSearchDelay": {
        "value": 1006,
        "options": {}
    },
    "routing": {
        "value": 1007,
        "options": {}
    },
    "insightsClient": {
        "value": 1008,
        "options": {}
    },
    "container": {
        "value": 1009,
        "options": {}
    },
    "placeholder": {
        "value": 1010,
        "options": {}
    },
    "autofocus": {
        "value": 1011,
        "options": {}
    },
    "searchAsYouType": {
        "value": 1012,
        "options": {}
    },
    "showReset": {
        "value": 1013,
        "options": {}
    },
    "showSubmit": {
        "value": 1014,
        "options": {}
    },
    "showLoadingIndicator": {
        "value": 1015,
        "options": {}
    },
    "queryHook": {
        "value": 1016,
        "options": {}
    },
    "templates": {
        "value": 1017,
        "options": {}
    },
    "cssClasses": {
        "value": 1018,
        "options": {}
    },
    "escapeHTML": {
        "value": 1019,
        "options": {}
    },
    "transformItems": {
        "value": 1020,
        "options": {}
    },
    "attributes": {
        "value": 1021,
        "options": {}
    },
    "showMore": {
        "value": 1022,
        "options": {}
    },
    "attribute": {
        "value": 1023,
        "options": {}
    },
    "escapeFacetValues": {
        "value": 1024,
        "options": {}
    },
    "searchParameters": {
        "value": 1025,
        "options": {}
    },
    "on": {
        "value": 1026,
        "options": {}
    },
    "items": {
        "value": 1027,
        "options": {}
    },
    "padding": {
        "value": 1028,
        "options": {}
    },
    "precision": {
        "value": 1029,
        "options": {}
    },
    "max": {
        "value": 1030,
        "options": {}
    },
    "searchable": {
        "value": 1031,
        "options": {}
    },
    "searchablePlaceholder": {
        "value": 1032,
        "options": {}
    },
    "searchableShowReset": {
        "value": 1033,
        "options": {}
    },
    "attributesToSnippet": {
        "value": 1034,
        "options": {}
    },
    "snippetEllipsisText": {
        "value": 1035,
        "options": {}
    },
    "removeWordsIfNoResults": {
        "value": 1036,
        "options": {}
    },
    "scrollTo": {
        "value": 1037,
        "options": {}
    },
    "showFirst": {
        "value": 1038,
        "options": {}
    },
    "showLast": {
        "value": 1039,
        "options": {}
    },
    "pips": {
        "value": 1040,
        "options": {}
    },
    "tooltips": {
        "value": 1041,
        "options": {}
    }
};

// Widget ========================================

var Widget = exports.Widget = {};

Widget.read = function (pbf, end) {
    return pbf.readFields(Widget._readField, {type: 0, params: [], useConnector: false}, end);
};
Widget._readField = function (tag, obj, pbf) {
    if (tag === 600) obj.type = pbf.readVarint();
    else if (tag === 601) pbf.readPackedVarint(obj.params);
    else if (tag === 602) obj.useConnector = pbf.readBoolean();
};
Widget.write = function (obj, pbf) {
    if (obj.type) pbf.writeVarintField(600, obj.type);
    if (obj.params) pbf.writePackedVarint(601, obj.params);
    if (obj.useConnector) pbf.writeBooleanField(602, obj.useConnector);
};

// Schema ========================================

var Schema = exports.Schema = {};

Schema.read = function (pbf, end) {
    return pbf.readFields(Schema._readField, {widgets: [], oversize: 0}, end);
};
Schema._readField = function (tag, obj, pbf) {
    if (tag === 501) obj.widgets.push(Widget.read(pbf, pbf.readVarint() + pbf.pos));
    else if (tag === 502) obj.oversize = pbf.readVarint(true);
};
Schema.write = function (obj, pbf) {
    if (obj.widgets) for (var i = 0; i < obj.widgets.length; i++) pbf.writeMessage(501, Widget.write, obj.widgets[i]);
    if (obj.oversize) pbf.writeVarintField(502, obj.oversize);
};
